Move Allocator and Device to ATen/core

jerryzh168 opened the pull request:
Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/10783

Differential Revision: D9466602

gchanan commented on aten/src/ATen/core/common.h:
```diff
@@ -0,0 +1,17 @@
+#pragma once
```
we already have ATenGeneral.h  / Macros.h for common stuff that is included, I don't think we need another file.  Maybe just move this to ATenGeneral.h?

gchanan commented on aten/src/ATen/core/Device.cpp:
```diff
@@ -64,7 +64,11 @@ Device::Device(const std::string& device_string) : Device(Type::CPU) {
   const auto index_string = device_string.substr(position);
   try {
+#if defined(__ANDROID__)
+    index_ = stoi(index_string);
```
is this how ANDROID-specific calls like this are handled in the caffe2 codebase?  It seems a bit strange because you'd have to guard every call site this way rather than have a single implementation that does the right thing, but I don't know much about mobile.

gchanan commented on aten/src/ATen/core/Macros.h:
```diff
@@ -29,3 +32,18 @@
+  return n;
+}
+#else
+using std::stoi;
```
this is not ideal, because now this function has different arguments depending on if you are on android or not.  Instead, how about just changing the body?

jerryzh168 replied:
This is what caffe2 currently do: https://github.com/pytorch/pytorch/blob/master/caffe2/core/common.h

do you mean
```
namespace at {
inline int stoi(const std::string& str) {
#if defined(__ANDROID__)
  std::stringstream ss;
  int n = 0;
  ss << str;
  ss >> n;
  return n;
#else
 return std::stoi(str);
#endif
}
} // namespace at
```
?

Then we'll probably need to change callsite to use at::stoi? is that OK?

gchanan replied:
that sounds good to me, what do you think?

jerryzh168 replied:
yeah, we can do that.

ezyang committed:
Move Allocator and Device to ATen/core

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/10798

Differential Revision: D9466602

fbshipit-source-id: 39c436c6269d2cff1c6f6911cd6d71311f6fd14f
